File: Point_to_Point/exercises/Fortran/solutions/ex1.f90
PROGRAM MAIN
    USE ISO_FORTRAN_ENV, ONLY : output_unit
    USE MPI
    IMPLICIT NONE
    CHARACTER*8 :: rank_string, nproc_string, nstring
    INTEGER :: i, ierr, mstatus, my_rank, num_proc
    INTEGER :: nnum=1
    REAL*8  :: the_dot_product = 0.0D0
    REAL*8, ALLOCATABLE :: your_numbers(:), my_numbers(:)
    CHARACTER*120 :: fprefix = 'Numbers_N', fname


    CALL INITIALIZATION()


    IF (my_rank .eq. 0) THEN

        OPEN(unit=15,file=fname,form='formatted', status='replace',access='stream')
        DO i = 1, nnum
            WRITE(15,'(i8,e12.3)')i, my_numbers(i)
        ENDDO
        CLOSE(15)

        ALLOCATE(your_numbers(1:nnum))
        your_numbers = 0
        CALL MPI_Recv(your_numbers, nnum, MPI_DOUBLE_PRECISION, 1, 1, MPI_COMM_WORLD, mstatus, ierr)
        
        CALL MPI_BARRIER(MPI_COMM_WORLD,ierr)
        DO i = 1, nnum
            the_dot_product = the_dot_product+your_numbers(i)*my_numbers(i)
        ENDDO

        WRITE(output_unit,*)'The product is: ', the_dot_product
    ELSE

        CALL MPI_Send(my_numbers,nnum,MPI_DOUBLE_PRECISION,0,my_rank,MPI_COMM_WORLD,ierr)
        CALL MPI_BARRIER(MPI_COMM_WORLD,ierr)

        OPEN(unit=15,file=fname,form='formatted', status='old',access='stream',position='append')
        DO i = 1, nnum
            WRITE(15,'(i,e12.3)')my_rank*nnum+i, my_numbers(i)
        ENDDO
        CLOSE(15)

    ENDIF


    CALL MPI_Finalize(ierr) 

    

CONTAINS


    SUBROUTINE Parse_Command_line(n)
            IMPLICIT NONE
            INTEGER, INTENT(INOUT)   :: n
            INTEGER :: nc                    ! Number of command-line arguments
            INTEGER :: i                    
            CHARACTER(len=1024) :: argname  ! Argument key
            CHARACTER(len=1024) :: val      ! Argument value

            nc = command_argument_count()
            DO i=1,nc,2
                    CALL get_command_argument(i, argname)
                    CALL get_command_argument(i+1, val)
                    SELECT CASE(argname)
                            CASE('-N')
                                    READ(val, '(I8)') n
                            CASE DEFAULT
                                    WRITE(output_unit,'(a)') ' '
                                    WRITE(output_unit,'(a)') &
                                    ' Unrecognized option: '// trim(argname)
                    END SELECT
            ENDDO

    END SUBROUTINE Parse_Command_Line


    SUBROUTINE Initialization()
        IMPLICIT NONE
        !//////////////////////////////
        ! Initialize MPI
        CALL MPI_INIT( ierr )
        CALL MPI_Comm_size(MPI_COMM_WORLD, num_proc,ierr)
        CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank,ierr)

        !/////////////////////////////////////////
        ! Rank 0 reads command-line arguments and 
        ! broadcasts to other ranks.
        IF (my_rank .eq. 0) CALL Parse_Command_Line(nnum)
        CALL MPI_Bcast(nnum,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)

        ! ~~~ Set the output filename ~~~
        WRITE(nstring,'(i8)')nnum
        fname = TRIM(fprefix)//ADJUSTL(TRIM(nstring))


        !/////////////////////////////////////////////
        ! Generate a sequence of N random numbers on 
        ! each process.
        ALLOCATE(my_numbers(1:nnum))
        CALL RANDOM_NUMBER(my_numbers)


    END SUBROUTINE Initialization

END PROGRAM MAIN

File: Point_to_Point/exercises/C++/solutions/ex1.cxx
#include <iostream>
#include <fstream>
#include <iomanip>
#include <cstdlib>
#include <cmath>
#include <random>

#include "mpi.h"

using namespace std;

int main ( int argc, char *argv[] );
int initialization(int& mr, int& np, int argc, char *argv[], double * &my_numbers);
		      
int main(int argc, char *argv[])
{
  double *your_numbers, *my_numbers;
  int my_rank, num_proc, ierr;
  
  int n = initialization(my_rank, num_proc, argc, argv, my_numbers);

  cout << "Process " << my_rank << " of " << num_proc << " Problem size: " << n << "\n";

  const string filename = string("Numbers_N")+string(argv[1]);

  // Processor zero does the following
  if (my_rank == 0) {
    ofstream myfile;
    myfile.open(filename, ios::out);
    for (int i = 0; i < n; i++) {
      myfile << my_numbers[i] << "\n";
    }
    myfile.close();
    // barrier here to make sure the file is created before the message exchange
    ierr = MPI_Barrier(MPI_COMM_WORLD);

    // Receive data from processor 1
    your_numbers = new double[n];
    ierr = MPI_Recv((void *) your_numbers, n, MPI_DOUBLE, 1, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

    // perform calculation
    double dprod = 0.0;
    for (int i=0; i<n; i++){
      dprod += my_numbers[i]*your_numbers[i];
    }
    cout << "The dot product is " << dprod << "\n";
  }

  // Processor 1 does the following
  if (my_rank == 1) {
    ierr = MPI_Barrier(MPI_COMM_WORLD);
    ierr = MPI_Send((void *) my_numbers, n, MPI_DOUBLE, 0, 1, MPI_COMM_WORLD);
    
    ofstream myfile;
    myfile.exceptions ( ofstream::failbit | ofstream::badbit );
    try {
      myfile.open(filename, ofstream::out | ofstream::app);
      for (int i = 0; i < n; i++) {
	myfile << my_numbers[i] << "\n";
      }
      myfile.close();
    }
    catch (const ofstream::failure& e) {
      cout << "Error appending to file\n";
    }
  }

  ierr = MPI_Finalize();						     
}

int initialization(int& my_rank, int& num_proc, int argc, char *argv[], double * &a)
{
  int ierr;
  int n;

  ierr = MPI_Init(&argc, &argv);
  ierr = MPI_Comm_size( MPI_COMM_WORLD, &num_proc);
  ierr = MPI_Comm_rank( MPI_COMM_WORLD, &my_rank);

  if ( my_rank == 0 ) {
      if ( 1 < argc ) {
	n = atoi ( argv[1] );
      }
  }

  ierr = MPI_Bcast(&n, 1, MPI_INT, 0, MPI_COMM_WORLD);
  a = new double[n];
  for (int i = 0; i < n; i++) {
    a[i] = (double) rand() / double (RAND_MAX);
  }
  cout << a[0] << n << "\n";
  return n;
}
